### PR TITLE
Auto create browser when disconnected

### LINKED ISSUE #366: Error: WebSocket is not open: readyState 3 (CLOSED)
Rendertron is working 24/7 and after some day (1-2) it fails with error
`Error: WebSocket is not open: readyState 3 (CLOSED)`
And browesr won't restart

This bug is [relative to Puppeteer](https://github.com/puppeteer/puppeteer/issues/4428) and it has [workaround](https://github.com/puppeteer/puppeteer/issues/4428#issuecomment-500184469)

```
  Error: WebSocket is not open: readyState 3 (CLOSED)
      at WebSocket.send (/home/user/rendertron/node_modules/ws/lib/websocket.js:329:19)
      at WebSocketTransport.send (/home/user/rendertron/node_modules/puppeteer/lib/WebSocketTransport.js:60:14)
      at Connection._rawSend (/home/user/rendertron/node_modules/puppeteer/lib/Connection.js:86:21)
      at Connection.send (/home/user/rendertron/node_modules/puppeteer/lib/Connection.js:72:21)
      at Browser._createPageInContext (/home/user/rendertron/node_modules/puppeteer/lib/Browser.js:174:47)
      at BrowserContext.newPage (/home/user/rendertron/node_modules/puppeteer/lib/Browser.js:367:26)
      at Browser.newPage (/home/user/rendertron/node_modules/puppeteer/lib/Browser.js:166:33)
      at Browser.<anonymous> (/home/user/rendertron/node_modules/puppeteer/lib/helper.js:112:23)
      at Renderer.serialize (/home/user/rendertron/src/renderer.ts:68:37)
      at Rendertron.handleRenderRequest (/home/user/rendertron/src/rendertron.ts:96:44)
    -- ASYNC --
      at Browser.<anonymous> (/home/user/rendertron/node_modules/puppeteer/lib/helper.js:111:15)
      at Renderer.serialize (/home/user/rendertron/src/renderer.ts:68:37)
      at Rendertron.handleRenderRequest (/home/user/rendertron/src/rendertron.ts:96:44)
      at /home/user/rendertron/node_modules/koa-route/index.js:39:44
      at dispatch (/home/user/rendertron/node_modules/koa-compose/index.js:42:32)
      at /home/user/rendertron/node_modules/koa-route/index.js:43:16
      at dispatch (/home/user/rendertron/node_modules/koa-compose/index.js:42:32)
      at /home/user/rendertron/node_modules/koa-route/index.js:43:16
      at dispatch (/home/user/rendertron/node_modules/koa-compose/index.js:42:32)
      at bodyParser (/home/user/rendertron/node_modules/koa-bodyparser/index.js:86:11)
```

### PR DESCRIPTION
**Pull request by abdonrd**
Fix #366

### CODE REVIEW

**Comment by AVGP** [comment]
Thanks for the PR!

There is a problem with just restarting the browser, unfortunately - the `Renderer` instance is connected to the browser instance (via CDP / web socket) and if we just create a new browser instance, our renderer will fail, because it no longer is connected to anything.

We need to also update the `Renderer` instance in order for this to work. I came up with the following piece of code for that, WDYT:

```typescript
  async createRenderer(config: Config, onRendererCreated: (renderer: Renderer) => void) {
    const browser = await puppeteer.launch({
      headless: false,
      args: ['--no-sandbox']
    });
    browser.on('disconnected', () => {
      this.createRenderer(config, onRendererCreated);
    });

    const renderer = new Renderer(browser, config);
    onRendererCreated(renderer);
  }
```
The reason for using a callback is that it makes the side effect of updating the `Renderer` instance a little more obvious, I think:

```typescript
    // The callback can be called multiple times, 
    // if the renderer needs to be updated(e.g.due to a browser crash)
    this.createRenderer(this.config, (renderer) => this.renderer = renderer);
```

**Comment by abdonrd** [comment]
Thanks! You're right, @AVGP!

Updating...

File: src/rendertron.ts [hunk 1]
```diff
@@ -22,15 +22,34 @@ export class Rendertron {
   private port = process.env.PORT || this.config.port;
   private host = process.env.HOST || this.config.host;
 
+  async createRenderer(
+    config: Config,
+    onRendererCreated: (renderer: Renderer) => void
+  ) {
+    const browser = await puppeteer.launch({
+      headless: false,
```
Comment (AVGP) [inline src/rendertron.ts]: Oops, this is from the test. Should be removed here

File: src/rendertron.ts [hunk 2]
```diff
@@ -22,15 +22,34 @@ export class Rendertron {
...
+      args: ['--no-sandbox']
+    });
+
+    browser.on('disconnected', () => {
+      this.createRenderer(config, onRendererCreated);
+    });
+
+    const renderer = new Renderer(browser, config);
```
Comment (AVGP) [inline src/rendertron.ts]: Should we just substitute the argument for the callback with this instead of this extra line. What do you think?

Comment (abdonrd) [reply src/rendertron.ts]: I think it is a bit simpler, yes!

**Comment by abdonrd** [comment]
Travis is happy now! 😄

**Pull request merged**